pkg/controller/direct/bigqueryanalyticshub/listing_controller.go

```diff
@@ -0,0 +1,336 @@
+		toProto := DataProvider_ToProto(mapCtx, a.desired.Spec.DataProvider)
+		if mapCtx.Err() != nil {
+			return fmt.Errorf("converting data provider: %w", mapCtx.Err())
+		}
+
+		if !reflect.DeepEqual(toProto, a.actual.DataProvider) {
+			updateMask.Paths = append(updateMask.Paths, "data_provider")
+		}
+	}
+
+	if a.desired.Spec.Publisher != nil {
+		mapCtx := &direct.MapContext{}
+		toProto := Publisher_ToProto(mapCtx, a.desired.Spec.Publisher)
+		if mapCtx.Err() != nil {
+			return fmt.Errorf("converting publisher: %w", mapCtx.Err())
+		}
+
+		if !reflect.DeepEqual(toProto, a.actual.Publisher) {
+			updateMask.Paths = append(updateMask.Paths, "publisher")
+		}
+	}
+
+	if a.desired.Spec.Categories != nil {
+		mapCtx := &direct.MapContext{}
+		toProto := Categories_ToProto(mapCtx, a.desired.Spec.Categories)
+		if mapCtx.Err() != nil {
+			return fmt.Errorf("converting categories: %w", mapCtx.Err())
+		}
+		if !reflect.DeepEqual(toProto, a.actual.Categories) {
+			updateMask.Paths = append(updateMask.Paths, "categories")
+		}
+	}
+
+	if len(updateMask.Paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id.External)
+		return nil
+	}
+
+	req := &bigqueryanalyticshubpb.UpdateListingRequest{
+		UpdateMask: updateMask,
+		Listing:    resource,
+	}
+	updated, err := a.gcpClient.UpdateListing(ctx, req)
+	if err != nil {
+		return fmt.Errorf("updating Listing %s: %w", a.id.External, err)
+	}
+
+	log.V(2).Info("successfully updated Listing", "name", a.id.External)
+
+	status := &krm.BigQueryAnalyticsHubListingStatus{}
+	status.ObservedState = BigQueryAnalyticsHubListingObservedState_FromProto(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+
+	return updateOp.UpdateStatus(ctx, status, nil)
+}
+
+func (a *ListingAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("Find() not called")
+	}
+	u := &unstructured.Unstructured{}
+
+	obj := &krm.BigQueryAnalyticsHubListing{}
+	mapCtx := &direct.MapContext{}
+	obj.Spec = direct.ValueOf(BigQueryAnalyticsHubListingSpec_FromProto(mapCtx, a.actual))
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+
+	parent, err := a.id.Parent()
+	if err != nil {
+		return nil, err
+	}
+	obj.Spec.ProjectRef = &refs.ProjectRef{External: parent.String()}
+	obj.Spec.Location = parent.Location
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+	u.Object = uObj
+	return u, nil
+}
+
+// Delete implements the Adapter interface.
+func (a *ListingAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx).WithName(listingCtrlName)
+	log.V(2).Info("deleting Listing", "name", a.id.External)
+
+	parent, err := a.id.Parent()
+	if err != nil {
+		return false, err
+	}
+
+	actualName := parent.String() + "/dataExchanges/" + a.references.dataExchange.DataExchangeID + "/listings/" + a.id.Name
```

Comment:
The parent should contain `dataExchanges/` so that we can validate the immutability in the api  <kind>_reference.go

Reply:
reworked the references for this resource